src/gnc_attitude_estimation.cpp

```diff
@@ -0,0 +1,123 @@
+//
+// src/gnc_attitude_estimationr.cpp
+// PSim
+//
+// Contributors:
+//   Kyle Krol  kpk63@cornell.edu
+//
+// Pathfinder for Autonomous Navigation
+// Space Systems Design Studio
+// Cornell Univeristy
+//
+
+#include <gnc_attitude_estimation.hpp>
+#include <gnc_environment.hpp>
+#include <gnc_utilities.hpp>
+
+#include <cmath>
+
+static_assert(std::numeric_limits<double>::has_quiet_NaN,
+    "GNC code requires quiet NaN's to be available.");
+static_assert(std::numeric_limits<float>::has_quiet_NaN,
+    "GNC code requires quiet NaN's to be available.");
+
+namespace gnc {
+
+AttitudeEstimatorState::AttitudeEstimatorState() {
+  this->t = std::numeric_limits<double>::quiet_NaN();
+  this->q_body_eci = lin::nans<decltype(this->q_body_eci)>();
+  this->w_buffer = CircularBuffer<lin::Vector3f, 5>();
+}
+
+AttitudeEstimatorData::AttitudeEstimatorData() {
+  this->t = std::numeric_limits<double>::quiet_NaN();
+  this->r_ecef = lin::nans<decltype(this->r_ecef)>();
+  this->b_body = lin::nans<decltype(this->b_body)>();
+  this->s_body = lin::nans<decltype(this->s_body)>();
+}
+
+AttitudeEstimate::AttitudeEstimate() {
+  this->q_body_eci = lin::nans<decltype(this->q_body_eci)>();
+  this->w_body = lin::nans<decltype(this->w_body)>();
+}
+
+/** @fn estimate_q_body_eci
+ *  @returns Zero of success.
+ *  Attempts to determine attitude given time, position, a sun vector, and
+ *  magnetic field vector. The result, if succesful, is written into the
+ *  quaternion q_body_eci. */
+static int estimate_q_body_eci(double t, lin::Vector3d const &r_ecef,
+    lin::Vector3f const &s_body, lin::Vector3f const &b_body, lin::Vector4f &q_body_eci) {
+  // Setup the rest of the triad inputs (including expected sun and magnetic
+  // field vectors in ECI)
+  lin::Vector3f s_eci, b_eci, r_ecef_f({
+    static_cast<float>(r_ecef(0)),
+    static_cast<float>(r_ecef(1)),
+    static_cast<float>(r_ecef(2))
+  });
+  env::sun_vector(t, s_eci);
+  env::magnetic_field(t, r_ecef_f, b_eci);
+  // Perform triad and return the result code (result is written to q_body_eci)
+  return utl::triad(s_eci, b_eci, s_body, b_body, q_body_eci);
```

Comment:
All the vectors going into triad need to be unit vectors, so try to normalize `b_body` and `b_eci`, fail the estimation if either are too small (under the noise/bias level of the magnetometer).

Reply:
Very good point! I totally missed that!